source/ranges.tex

```diff
@@ -12881,3 +12893,761 @@
+      requires @\exposconcept{cartesian-product-is-random-access}@<Const, First, Vs...>;
+
+    constexpr reference operator[](difference_type n) const
+      requires @\exposconcept{cartesian-product-is-random-access}@<Const, First, Vs...>;
+
+    friend constexpr bool operator==(const @\exposid{iterator}@& x, const @\exposid{iterator}@& y)
+      requires @\libconcept{equality_comparable}@<iterator_t<@\exposid{maybe-const}@<Const, First>>>;
+
+    friend constexpr bool operator==(const @\exposid{iterator}@& x, default_sentinel_t);
+
+    friend constexpr auto operator<=>(const @\exposid{iterator}@& x, const @\exposid{iterator}@& y)
+      requires @\exposconcept{all-random-access}@<Const, First, Vs...>;
+
+    friend constexpr @\exposid{iterator}@ operator+(const iterator& x, difference_type y)
+      requires @\exposconcept{cartesian-product-is-random-access}@<Const, First, Vs...>;
+    friend constexpr @\exposid{iterator}@ operator+(difference_type x, const iterator& y)
+      requires @\exposconcept{cartesian-product-is-random-access}@<Const, First, Vs...>;
+    friend constexpr iterator operator-(const iterator& x, difference_type y)
+      requires @\exposconcept{cartesian-product-is-random-access}@<Const, First, Vs...>;
+    friend constexpr difference_type operator-(const iterator& x, const iterator& y)
+      requires cartesian-is-sized-sentinel<Const, iterator_t, First, Vs...>;
+
+    friend constexpr difference_type operator-(iterator i, default_sentinel_t)
+      requires cartesian-is-sized-sentinel<Const, sentinel_t, First, Vs...>;
+    friend constexpr difference_type operator-(default_sentinel_t, iterator i)
+      requires @\exposconcept{cartesian-is-sized-sentinel}@<Const, sentinel_t, First, Vs...>;
+
+    friend constexpr auto iter_move(const @\exposid{iterator}@& i) noexcept(@\seebelow@);
+
+    friend constexpr void iter_swap(const @\exposid{iterator}@& l, const @\exposid{iterator}@& r) noexcept(@\seebelow@)
+      requires (@\libconcept{indirectly_swappable}@<iterator_t<@\exposid{maybe-const}@<Const, First>>> && ... &&
+        @\libconcept{indirectly_swappable}@<iterator_t<@\exposid{maybe-const}@<Const, Vs>>>);
+
+  private:
+    @\exposid{maybe-const}@<Const, cartesian_product_view>* @\exposid{parent_}@ = nullptr; // \expos
+    @\exposid{tuple-or-pair}@<iterator_t<@\exposid{maybe-const}@<Const, First>>,
+      iterator_t<@\exposid{maybe-const}@<Const, Vs>>...> @\exposid{current_}@; // \expos
+
+    template<size_t N = sizeof...(Vs)>
+      constexpr void @\exposid{next}@(); // \expos
+
+    template<size_t N = sizeof...(Vs)>
+      constexpr void @\exposid{prev}@(); // \expos
+
+    template<class Tuple>
+      constexpr difference_type @\exposid{distance-from}@(Tuple t); // \expos
+
+    constexpr explicit @\exposid{iterator}@(@\exposid{tuple-or-pair}@<iterator_t<@\exposid{maybe-const}@<Const, First>>,
+      iterator_t<@\exposid{maybe-const}@<Const, Vs>>...> current); // \expos
```

Comment:
Why should we deviate fromi the incoming paper? Because we can?

Reply:
From what I know, the less deviations, the less exceptions the Editor has to call out in the Editor's Report. There's also the cost of determining whether something's editorial.

Unfortunately, the redundancy here brought about https://github.com/cplusplus/draft/pull/5647/files#r929145700. I thought this was fair, given that it is a `private` constructor and the lengthy type is the same as `decltype(`_`current_`_`)`.

This meeting brought along with it a lot to review. I know it's a once in a 3 years event. But I'm going to change my approach to be less time consuming for me and only review what has actually been committed to `main`. That way, I can open PRs for obvious editorial commits and issues that won't get buried between many unrelated comments.

Reply:
@JohelEGP , you reviews are invaluable. It would be a pity if we'd apply all those `\exposid`s etc. only after the merge, instead of just applying the correct patch from the onset.

I understand there are occasionally issues that we'd like to postpone, and maybe we need to do a better job at tracking and addressing them.

Reply:
> It would be a pity if we'd apply all those `\exposid`s etc. only after the merge, instead of just applying the correct patch from the onset.

I feel you. I think it may be that I'm just too eager to review and end up taking on too much. Perhaps it'd be better to delay some until others have had a chance. But the amount to review stays the same, even if it's all fixed up when I come to it.

> and maybe we need to do a better job at tracking and addressing them.

We already open issues pointing to the relevant comments when that happens, so it's all good.

I used to try and contact authors to do this even earlier. That's too unmanageable, so I settled for setting trench on these motion PRs.